coffea/jetmet_tools/CorrectedJetsFactory.py

```diff
@@ -0,0 +1,351 @@
+                                   min_jet_pt_corr,
+                                   smearfact)
+
+        def getfunction(layout, depth):
+            if (isinstance(layout, awkward1.layout.NumpyArray)
+                or not isinstance(layout, (awkward1.layout.Content,
+                                           awkward1.partition.PartitionedArray)
+               )):
+                return lambda: awkward1.layout.NumpyArray(smearfact)
+            return None
+        smearfact = awkward1._util.recursively_apply(
+            awkward1.operations.convert.to_layout(jetPt),
+            getfunction)
+        smearfact = awkward1._util.wrap(smearfact, awkward1._util.behaviorof(jetPt))
+    else:
+        raise Exception('\'arr\' must be an awkward array of some kind!')
+
+    return smearfact
+
+
+class CorrectedJetsFactory(object):
+
+    def __init__(self, name_map, jec_stack):
+        # from PhysicsTools/PatUtils/interface/SmearedJetProducerT.h#L283
+        self.forceStochastic = False
+
+        if 'ptRaw' not in name_map or name_map['ptRaw'] is None:
+            warnings.warn('There is no name mapping for ptRaw,'
+                          ' CorrectedJets will assume that <object>.pt is raw pt!')
+            name_map['ptRaw'] = name_map['JetPt'] + '_raw'
+        self.treat_pt_as_raw = 'ptRaw' not in name_map
+
+        if 'massRaw' not in name_map or name_map['massRaw'] is None:
+            warnings.warn('There is no name mapping for massRaw,'
+                          ' CorrectedJets will assume that <object>.mass is raw pt!')
+            name_map['ptRaw'] = name_map['JetMass'] + '_raw'
+
+        total_signature = set()
+        for part in _stack_parts:
+            attr = getattr(jec_stack, part)
+            if part is not None:
+                total_signature.update(attr.signature)
+
+        missing = total_signature - set(name_map.keys())
+        if len(missing) > 0:
+            raise Exception(f'Missing mapping of {missing} in name_map!' +
+                            ' Cannot evaluate jet corrections!' +
+                            ' Please supply mappings for these variables!')
+
+        if 'ptGenJet' not in name_map:
+            warnings.warn('Input JaggedCandidateArray must have "ptGenJet" in order to apply hybrid JER smearing method. Stochastic smearing will be applied.')
+            self.forceStochastic = True
+
+        self.real_sig = [v for k, v in name_map.items()]
+        self.name_map = name_map
+        self.jec_stack = jec_stack
+
+    def uncertainties(self):
+        out = ['JER'] if self.jec_stack.jer is not None else []
+        if self.jec_stack.junc is not None:
+            out.extend(['JES_{0}'.format(unc) for unc in self.jec_stack.junc.levels])
+        return out
+
+    def build(self, jets, lazy_cache):
+        if lazy_cache is None:
+            raise Exception('CorrectedJetsFactory requires a awkward-array cache to function correctly.')
```

Comment:
For NanoEvents, one can likely assume that `jets._caches[0]` is a good lazy cache. (e.g. check `list(events.Jet._caches[0].keys())` out) I realize now that `array.cache` disappeared as a public attribute, so now this is private. Another option is to check for `jets.behavior["__events_factory__"]._cache` which is also private but coffea-private rather than awkward-private.

Reply:
The magic `__events_factory__` key is used by NanoEvents mixins to be able to resolve cross-references even from sliced arrays or records, so it will stay around for a while. In most cases its `_cache` attribute is the same runtime cache passed to `ak.from_arrayset`